stock_generate_putaway_from_inventory/models/stock_inventory.py

```diff
@@ -8,52 +8,51 @@
 class StockInventory(models.Model):
     _inherit = "stock.inventory"
 
-    def action_generate_putaway_strategy(self):
-        for inventory in self:
-            inventory._generate_putaway_strategy()
-
-    def _get_putaway_strategy(self, loc):
-        if loc.putaway_strategy_id:
-            return loc.putaway_strategy_id
-        elif loc.location_id:
-            return self._get_putaway_strategy(loc.location_id)
-
-    def _update_product_putaway_strategy(self, inventory_line, strategy):
-        putaway_line_obj = self.env["stock.fixed.putaway.strat"]
-        putaway_line = putaway_line_obj.search(
+    def action_generate_putaway_rules(self):
+        self._generate_putaway_rules()
+
+    def _generate_putaway_rules(self):
+        for record in self:
+            if record.state != "done":
+                raise ValidationError(
+                    _(
+                        "Please validate the inventory before generating "
+                        "the putaway strategy."
+                    )
+                )
+            for location in record.location_ids:
+                record.line_ids._generate_putaway_rules(location)
+
+
+class StockInventoryLine(models.Model):
+    _inherit = "stock.inventory.line"
+
+    def _generate_putaway_rules(self, inventory_location):
+        # Eliminate lines for other IN locations
+        # and eliminate lines where quantity is null
+        self.filtered(
+            lambda x: x.location_id._is_child(inventory_location) and x.product_qty > 0
+        )._update_product_putaway_rule(inventory_location)
```

Comment:
I think it looks better and maybe faster:
<pre>
def _generate_putaway_rules(self, inventory_locations):
    for location_in in inventory_locations:
        self.filtered(
            lambda x: x.product_qty &gt; 0.0
                and x.parent_path.startswith((location_in.parent_path or '') + '/')
        )._update_product_putaway_rule(location_in)
</pre>